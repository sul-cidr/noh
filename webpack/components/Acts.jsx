import React, { Component } from "react";
import PropTypes from "prop-types";

class Acts extends Component {
  createActs() {
    const length = this.props.acts.length - 1;
    let left = 0;
    return this.props.acts.map((act, index) => {
      const width = (100 * act.duration) / this.props.duration;
      const transcription = act.transcription ? (
        <span className="transcription">{act.transcription}</span>
      ) : (
        ""
      );
      left += width;
      const actEnd =
        index < length ? (
          <div
            key={`act-end-${act.translation}-${width}`}
            className="act__first-end"
            style={{ left: `${left}%` }}
          />
        ) : (
          ""
        );
      return [
        <div
          key={`act-${act.translation}-${width}`}
          className="act"
          style={{ width: `${width}%` }}
        >
          {transcription}
        </div>,
        actEnd
      ];
    });
  }

  render() {
    const acts = this.createActs();
    return <div className="acts-container">{acts}</div>;
  }
}

Acts.propTypes = {
  acts: PropTypes.arrayOf(
    PropTypes.shape({
      translation: PropTypes.string.isRequired,
      transcription: PropTypes.string.isRequired,
      duration: PropTypes.number.isRequired
    })
  ).isRequired,
  duration: PropTypes.number.isRequired
};

export default Acts;
