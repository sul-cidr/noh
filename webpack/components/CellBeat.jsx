import React from "react";
import PropTypes from "prop-types";

const CellBeat = props => (
  <div
    className={`cell cell--${props.length} cell--${
      props.beatText.toLowerCase() === "unmetered" ? "unmetered" : "beat"
    }`}
  >
    {props.beatText}
  </div>
);

CellBeat.propTypes = {
  beatText: PropTypes.string.isRequired,
  length: PropTypes.number
};

CellBeat.defaultProps = {
  length: 1
};

export default CellBeat;
