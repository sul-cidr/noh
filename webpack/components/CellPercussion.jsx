import React from "react";
import PropTypes from "prop-types";

const CellPercussion = props => {
  const percussionClass = props.text.startsWith("#")
    ? `cell--percussion-${props.text.slice(1)}`
    : "";
  const percussionContent = props.text.startsWith("#") ? "" : props.text;
  return (
    <div
      className={`cell cell--${props.length} cell--percussion ${percussionClass}`}
    >
      {percussionContent}
    </div>
  );
};

CellPercussion.propTypes = {
  text: PropTypes.string.isRequired,
  length: PropTypes.number.isRequired
};

export default CellPercussion;
