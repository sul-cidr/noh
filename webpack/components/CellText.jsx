import React from "react";
import PropTypes from "prop-types";

const CellText = props => {
  let rangeSpan = null;
  let rangeClass = "";
  if (props.vocalRange === "l") {
    rangeClass = "low";
  } else if (props.vocalRange === "m") {
    rangeClass = "medium";
  } else if (props.vocalRange === "h") {
    rangeClass = "high";
  }
  if (props.vocalRange !== "") {
    rangeSpan = <span className={`cell__range--${rangeClass}`} />;
  } else {
    rangeSpan = null;
  }
  let typeSpan = null;
  let tooltipSpan = null;
  if (props.voiceType !== "") {
    const beginClass = props.beginning === true ? "cell__character--begin" : "";
    const endClass = props.end === true ? "cell__character--end" : "";
    typeSpan = (
      <span
        className={`cell__character cell__character--${
          props.voiceType
        } ${beginClass} ${endClass}`}
      />
    );
    tooltipSpan = <span className="cell__tooltip">{props.voiceType}</span>;
  }
  return (
    <div className={`cell cell--${props.length} cell--text`}>
      <span
        title={props.text}
        className={`truncate ${props.voiceType !== "" ? "voice__label" : ""}`}
      >
        {props.text}
      </span>
      {typeSpan}
      {tooltipSpan}
      {rangeSpan}
    </div>
  );
};

CellText.propTypes = {
  text: PropTypes.string.isRequired,
  length: PropTypes.number.isRequired,
  vocalRange: PropTypes.string,
  voiceType: PropTypes.string,
  beginning: PropTypes.bool,
  end: PropTypes.bool
};

CellText.defaultProps = {
  vocalRange: "",
  voiceType: "",
  beginning: false,
  end: false
};

export default CellText;
