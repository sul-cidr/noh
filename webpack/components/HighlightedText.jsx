import React from "react";
import PropTypes from "prop-types";
import TextLine from "./TextLine";

// containing component needs to know about currentTime and
// calculate currentPhraseID to pass into this component
class HighlightedText extends React.Component {
  createLines() {
    return this.props.phrases.map(phrase => (
      <TextLine
        key={phrase.phraseID}
        translation={phrase.translation}
        transcription={phrase.transcription}
        active={phrase.phraseID === this.props.currentPhraseID}
      />
    ));
  }

  render() {
    const lines = this.createLines();
    return (
      <div className="transcription">
        <div className="transcription__title">
          <p>
            Singing style: <span>{this.props.singingStyle}</span>
          </p>
        </div>
        <div className="transcription__text">{lines}</div>
      </div>
    );
  }
}

HighlightedText.propTypes = {
  singingStyle: PropTypes.string.isRequired,
  currentPhraseID: PropTypes.string.isRequired,
  phrases: PropTypes.arrayOf(
    PropTypes.shape({
      phraseID: PropTypes.string,
      startTime: PropTypes.number,
      endTime: PropTypes.number,
      transcription: PropTypes.string,
      translation: PropTypes.string
    })
  ).isRequired
};

export default HighlightedText;
