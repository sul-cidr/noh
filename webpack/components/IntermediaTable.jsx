import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import IntermediaTitle from "./IntermediaTitle";
import IntermediaElement from "./IntermediaElement";

class IntermediaTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentSectionIndex: 0
    };
  }

  componentWillReceiveProps() {
    // Whenever props change, e.g. time, check to see if the current section index has changed. If so, update state.
    const currentSectionIndex = this.findCurrentSectionIndex();
    if (currentSectionIndex !== this.state.currentSectionIndex) {
      this.updateIndex(currentSectionIndex);
    }
  }

  findCurrentSectionIndex() {
    // Determine current sectionIndex based on time
    const currentSectionIndex = this.props.sections.findIndex(
      section =>
        this.props.currentTime >= section.startTime.value &&
        this.props.currentTime < section.endTime.value
    );
    // since findIndex returns -1 if it finds nothing, just return the current section index from state
    if (currentSectionIndex === -1) {
      return this.state.currentSectionIndex;
    }
    return currentSectionIndex;
  }

  updateIndex(currentSectionIndex) {
    this.setState({ currentSectionIndex });
  }

  render() {
    // due to lack of data for start and end time, not always finding current section
    const section = this.props.sections[this.state.currentSectionIndex];
    return (
      <div className="intermedia-table">
        <IntermediaTitle
          section={section.sectionName.value}
          play={this.props.play}
        />
        <IntermediaElement
          fieldName="Voices"
          fieldValue={section.numberVoices.value}
        />
        <IntermediaElement
          fieldName="Type of voice"
          fieldValue={section.voice.value}
        />
        <IntermediaElement fieldName="Text" fieldValue={section.text.value} />
        <IntermediaElement
          fieldName="Percussion"
          fieldValue={section.percussion.value}
        />
        <IntermediaElement
          fieldName="Type of percussion"
          fieldValue={section.numberOfPercussion.value}
        />
        <IntermediaElement
          fieldName="Nohkan"
          fieldValue={section.nokhanPresent.value}
        />
        <IntermediaElement
          fieldName="Dance"
          fieldValue={section.dancePresent.value}
        />
      </div>
    );
  }
}

IntermediaTable.propTypes = {
  play: PropTypes.string.isRequired,
  currentTime: PropTypes.number.isRequired,
  sections: PropTypes.arrayOf(
    PropTypes.shape({
      play: PropTypes.shape({ value: PropTypes.string }),
      sectionName: PropTypes.shape({ value: PropTypes.string }),
      intensity: PropTypes.shape({ number: PropTypes.string }),
      startTime: PropTypes.shape({ value: PropTypes.number }),
      endTime: PropTypes.shape({ value: PropTypes.number }),
      numberVoices: PropTypes.shape({
        number: PropTypes.string,
        value: PropTypes.string
      }),
      voice: PropTypes.shape({ value: PropTypes.string }),
      text: PropTypes.shape({ value: PropTypes.string }),
      numberOfPercussion: PropTypes.shape({ number: PropTypes.string }),
      percussion: PropTypes.shape({ value: PropTypes.string }),
      nokhanPresent: PropTypes.shape({
        present: PropTypes.string,
        value: PropTypes.string
      }),
      dancePresent: PropTypes.shape({
        present: PropTypes.string,
        value: PropTypes.string
      }),
      captions: PropTypes.Array,
      narrative: PropTypes.shape({ value: PropTypes.string })
    })
  ).isRequired
};

const mapStateToProps = state => ({ currentTime: state.currentTime });

export const Unwrapped = IntermediaTable;
export default connect(mapStateToProps)(IntermediaTable);
