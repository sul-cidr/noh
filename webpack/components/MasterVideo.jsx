/* eslint-disable jsx-a11y/media-has-caption */
import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { setCurrentTime, setIsPlaying } from "../actionCreators";

class MasterVideo extends Component {
  constructor(props) {
    super(props);

    this.video = null;
  }

  componentDidMount() {
    this.forceUpdate();
  }

  componentDidUpdate() {
    if (this.video.currentTime !== this.props.currentTime) {
      this.video.currentTime = this.props.currentTime;
    }
  }

  createTracks() {
    return this.props.tracks.map(track => (
      <track
        key={track.label}
        label={track.label}
        kind={track.kind}
        srcLang={track.lang}
        src={track.url}
      />
    ));
  }

  render() {
    const tracks = this.createTracks();
    return (
      <div>
        <video
          id="player"
          ref={video => {
            this.video = video;
          }}
          controls
          onTimeUpdate={event => this.props.updateCurrentTime(event)}
          onPlaying={this.props.updateIsPlaying}
          onEnded={this.props.updateIsNotPlaying}
          onPlay={this.props.updateIsNotPlaying}
          onPause={this.props.updateIsNotPlaying}
          onStalled={this.props.updateIsNotPlaying}
          onSeeking={this.props.updateIsNotPlaying}
          onWaiting={this.props.updateIsNotPlaying}
        >
          <source src={this.props.videoUrl} type="video/mp4" />
          {tracks}
        </video>
        <h3>{this.props.currentTime}</h3>
        <span>{this.props.startTime}</span>
      </div>
    );
  }
}

MasterVideo.propTypes = {
  currentTime: PropTypes.number,
  videoUrl: PropTypes.string,
  startTime: PropTypes.number,
  updateCurrentTime: PropTypes.func,
  updateIsPlaying: PropTypes.func,
  updateIsNotPlaying: PropTypes.func,
  tracks: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      kind: PropTypes.string,
      lang: PropTypes.string,
      url: PropTypes.string
    })
  )
};

MasterVideo.defaultProps = {
  currentTime: 0,
  videoUrl: "",
  startTime: 0,
  updateCurrentTime: () => {},
  updateIsPlaying: () => {},
  updateIsNotPlaying: () => {},
  tracks: []
};

const mapStateToProps = state => ({
  currentTime: state.currentTime,
  isPlaying: state.isPlaying,
  startTime: state.startTime
});

// The last two functions should be refactored
const mapDispatchToProps = dispatch => ({
  updateCurrentTime: event => {
    dispatch(setCurrentTime(event.target.currentTime));
  },
  updateIsPlaying: () => {
    dispatch(setIsPlaying(true));
  },
  updateIsNotPlaying: () => {
    dispatch(setIsPlaying(false));
  }
});

export const Unwrapped = MasterVideo;
export default connect(mapStateToProps, mapDispatchToProps)(MasterVideo);
