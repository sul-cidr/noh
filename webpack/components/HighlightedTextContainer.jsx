import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import HighlightedText from "./HighlightedText";
import { setCurrentPhraseID } from "../actionCreators";

class HighlightedTextContainer extends React.Component {
  componentWillReceiveProps() {
    const currentPhraseID = this.determineCurrentPhrase();
    this.props.setCurrentPhrase(currentPhraseID);
  }

  determineCurrentPhrase() {
    let currentPhraseID = "";
    for (let i = 0; i < this.props.phrases.length; i += 1) {
      if (
        this.props.currentTime >= this.props.phrases[i].startTime &&
        this.props.currentTime <= this.props.phrases[i].endTime
      ) {
        currentPhraseID = this.props.phrases[i].phraseID;
      }
    }
    return currentPhraseID;
  }

  render() {
    return (
      <div>
        <HighlightedText
          singingStyle={this.props.singingStyle}
          phrases={this.props.phrases}
          currentPhraseID={this.props.currentPhraseID}
        />
      </div>
    );
  }
}

HighlightedTextContainer.propTypes = {
  currentTime: PropTypes.number.isRequired,
  singingStyle: PropTypes.string.isRequired,
  currentPhraseID: PropTypes.string.isRequired,
  phrases: PropTypes.arrayOf(
    PropTypes.shape({
      phraseID: PropTypes.string,
      startTime: PropTypes.number,
      endTime: PropTypes.number,
      transcription: PropTypes.string,
      translation: PropTypes.string
    })
  ).isRequired,
  setCurrentPhrase: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  currentTime: state.currentTime,
  currentPhraseID: state.currentPhraseID
});

const mapDispatchToProps = dispatch => ({
  setCurrentPhrase: idString => {
    dispatch(setCurrentPhraseID(idString));
  }
});

export const Unwrapped = HighlightedTextContainer;
export default connect(mapStateToProps, mapDispatchToProps)(
  HighlightedTextContainer
);
