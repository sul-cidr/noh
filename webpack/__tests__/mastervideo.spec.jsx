import React from "react";
import { mount } from "enzyme";
import { Unwrapped as UnwrappedMasterVideo } from "../components/MasterVideo";

describe("<MasterVideo>", () => {
  it("renders as expected", () => {
    const container = document.createElement("div");
    const component = mount(<UnwrappedMasterVideo />, {
      attachTo: container
    });
    expect(component).toMatchSnapshot();
    expect(component.instance().video).not.toBe(null);
    expect(component.instance().video).not.toBe(undefined);
  });

  it("updates video currentTime when currentTime prop is set", () => {
    const container = document.createElement("div");
    const component = mount(<UnwrappedMasterVideo currentTime={10} />, {
      attachTo: container
    });
    component.setProps({ currentTime: 40 });
    expect(component.instance().video.currentTime).toBe(40);
  });
});
