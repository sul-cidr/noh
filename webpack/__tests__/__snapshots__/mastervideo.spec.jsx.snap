// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<MasterVideo> renders as expected 1`] = `
<MasterVideo
  currentTime={0}
  currentTimeOrigin=""
  tracks={[]}
  updateCurrentTime={null}
  videoUrl=""
>
  <div>
    <video
      controls={true}
      controlsList="nodownload"
      id="player"
      onSeeked={[Function]}
      onSeeking={[Function]}
      onTimeUpdate={[Function]}
    >
      <source
        src=""
        type="video/mp4"
      />
    </video>
  </div>
</MasterVideo>
`;
